airflow/utils/db.py

```diff
@@ -58,22 +59,28 @@
 
 log = logging.getLogger(__name__)
 
+SESSION_MUST_BE_ASSIGNED = "The session should be assigned here. This is a critical assertion."
+
 
 def _format_airflow_moved_table_name(source_table, version):
     return "__".join([settings.AIRFLOW_MOVED_TABLE_PREFIX, version.replace(".", "_"), source_table])
 
 
 @provide_session
-def merge_conn(conn, session=None):
+def merge_conn(conn, session: Optional[SASession] = None):
     """Add new Connection."""
+    if not session:
+        raise RuntimeError(SESSION_MUST_BE_ASSIGNED)
     if not session.query(Connection).filter(Connection.conn_id == conn.conn_id).first():
         session.add(conn)
         session.commit()
 
 
 @provide_session
-def add_default_pool_if_not_exists(session=None):
+def add_default_pool_if_not_exists(session: Optional[SASession] = None):
     """Add default pool if it does not exist."""
+    if not session:
+        raise RuntimeError(SESSION_MUST_BE_ASSIGNED)
```

Comment:
@uranusjr has a better plan for this that doesn't have a runtime impact.

Reply:
Which has now been merged to main.

Reply:
Corrected.